G1128 & Library Upgrades

nvasta opened the pull request:
First of all this branch is a bit old so we might want to pull master into it before merging the changes.
The changes themselves involve multiple aspects including:

- Update of the Copyright fields (to 2025)
- Update of G1128 to match the latest specification
- The new G1128 library being used contains both versions 1.3 and 1.7 (the latest)
- Moving the G1128Utils inside the G1128 library
- Fixes in the order of the Flyway migrations, including a full schema in the first migration, hopefully that doesn't cause issues. The G1128 migration is java based and therefore place inside a java package (db.migration)
- Porting of old G1128 1.3 XMLs into the 1.7 equivalent, but note that certain fields might be left empty
- Update of the Springboot version to 3.5.0
- Update of the jQuery Datatables library used in 2.3.1. There if a [bug](https://datatables.net/forums/discussion/78278/search-panes-sorting-the-panel-content-stop-working-at-the-third-click) which appears if you sort a column three times in a row, but we are working on it
- Additional GUI components to handle the multiple design and service specifications implemented by a service
- Additional database fields for the Instance class (statusEndpointUri, serviceTypes) and support in the DTO

Unfortunately, a big one to review so take your time!

Commit:
Updating the G1128 version to 1.7

Commit:
Changing the naming converion for the service types and Springboot upgrade

Commit:
Adding migrations for porting

Commit:
Using the latest G1128 library with migration capabilities

Commit:
Adding migration but we are missing fields in the porting

Commit:
Adding a try catch clause

Commit:
Copyright updates

Commit:
Adding initial db migration and upgrading the datatables

Commit:
Fixing the broken build

Commit:
Moving towards multiple design and service specs in the front end

Commit:
Fixing the table behaviour

Commit:
Geometry fixes

Commit:
Finalising the fixes

Commit:
Disabling the service type and data type selected on G1128

nvasta committed:
Merge pull request #46 from maritimeconnectivity/master

Merging changes from MCC fork

Commit:
Removing commented out code

oliverhaagh commented:
Awesome, thanks @nvasta! :D 
As you also suggested, I will start by pulling in the changes from master which should hopefully make the review a bit easier. :)

nvasta commented:
Might be a good idea to run the migration in a copy of the live database first to see if the porting works as expected. I couldn't do it myself, so I only run partial examples.

oliverhaagh commented:
Yes, I will try to do that. I can see that you made changes to the initial migration script which in my experience is something that Flyway doesn't really like...

nvasta commented:
Yes, indeed... but I think because it is already on V2 of the migrations it might let it through...

oliverhaagh commented on src/main/resources/db/migration/V1__Initial.sql:
Haven't tested it yet, but Flyway usually doesn't like changing existing migrations.

oliverhaagh commented on src/main/java/net/maritimeconnectivity/serviceregistry/models/dto/InstanceDto.java:
```diff
@@ -64,18 +62,18 @@ public class InstanceDto implements Serializable, JsonSerializable {
     private Set<Long> docIds = new HashSet<>();
-    private String implementsServiceDesign;
-    private String implementsServiceDesignVersion;
+    private Map<String, String> implementsServiceDesigns;
+    private Map<String, String> designsServiceSpecifications;
```
Can see that these are also defined as Maps in the Instance class, but I can't figure out what they are supposed to map from and to. Would it be possible to use a (Hash)Set instead?

nvasta replied:
Map between MRN of design/service-spec and version

oliverhaagh reviewed (commented):
A few initial comments

oliverhaagh commented on src/main/resources/application.yaml:
```diff
@@ -0,0 +1,125 @@
+        password: ${service.variable.datasource.database.password}
+    flyway:
+        enabled: true
+        url: jdbc:${service.variable.datasource.server.type}://${service.variable.datasource.server.host}:${service.variable.datasource.server.port}/${service.variable.datasource.database.name}
+        schemas: mcp_service_registry
+        user: ${service.variable.datasource.database.username}
+        password: ${service.variable.datasource.database.password}
```
```suggestion
```
Flyway will by default use the datasource that is defined above

nvasta replied:
OK will remove that part

Commit:
Polishing up for pull request comments

oliverhaagh commented on src/main/resources/application.yaml:
```diff
@@ -28,12 +28,6 @@ spring:
         username: ${service.variable.datasource.database.username}
         password: ${service.variable.datasource.database.password}
-    flyway:
-        enabled: true
```
I think these two lines still need to be there to enable flyway

oliverhaagh replied:
Can see that we actually also still need the line with `schemas: mcp_service_registry` as flyway seems to not like not having it there

Commit:
Adding back flyway

oliverhaagh commented:
As I was afraid it seems that Flyway does not like changing existing migrations as it fails with an error message starting with this:
```
Migration checksum mismatch for migration version 1
-> Applied to database : -1627021776
-> Resolved locally    : -1712307779
```

oliverhaagh commented:
What I think could maybe work is manually deleting the original flyway history in the database before running the migration. Also, because we already have setup a database structure, the V1 script will need to use `CREATE TABLE IF NOT EXISTS` instead of `CREATE TABLE` and the same for `SEQUENCE`. 
Apart from that I think any lines that sets the `OWNER` of different resources need to be deleted, as they are being set to the user `admin` which is not necessarily the same as the user being set for the datasource.

nvasta commented:
There is another migration V2 which has already been played... not sure if it can also be replayed.

I do get the point for the OWNER though... I will remove those statements.

oliverhaagh commented:
I just tried removing the OWNER statements and now flyway is complaining that it cannot change the ownership...

nvasta commented:
Errr?

oliverhaagh commented:
Yep...: 
```
org.postgresql.util.PSQLException: ERROR: cannot change ownership of identity sequence
  Detail: Sequence "document_id_seq" is linked to table "document".
```

Commit:
Removing the owner from the initial migration script

nvasta commented:
OK I think I got it... so the flyway scheme should be public... or at least that's how it's setup in the production database. I also removed the owner from the initial migration.
Now all you need to do it fix the chesum to: 436279054
BTW the java-based G1128 XML porting works but it does not create a checksum!

nvasta commented:
Oh also after the upgrade, the management portal might not work with the MSR... not sure about that...

oliverhaagh commented:
Finally got around to look at this again! The migration seems to work fine after manually fixing the checksum for the first version. For the java-based migration we could maybe consider setting the checksum manually by implementing the `getChecksum` function.

Commit:
Added comments and checksum to java migration

nvasta commented:
Sorry for the delay. I've now added a checksum function to the Java migration. I just used a fixed UID value... not sure what's the best way to do it.

Also added some comments :).

Let me know if you have any more comments!

oliverhaagh commented:
I think that should be fine, don't think we need to do anything fancy :) 
Unless there are some last minute changes you want to do, I think we can just go ahead and merge :)
